File: src/data/wuhen.js
'use strict';

// 无痕中转（api.wuhen-ai.com）— 余额 + 用量查询，仿 deepseek.js 范式。
// key 从 ~/.zcode/v2/config.json 按 baseURL 关键字匹配（无痕gpt/无痕grok）；
// GET /v1/usage 一次拿余额 + 今日/累计 + 按日 + 按模型。后台刷新 + 缓存。

const https = require('https');
const { providerApiKey } = require('./volc');

const USAGE_URL = 'https://api.wuhen-ai.com/v1/usage';
const CACHE_TTL = 15.0;

let cache = { ts: 0, payload: null };

function getWuhenUsage() {
  if (cache.payload) return cache.payload;
  return {
    enabled: Boolean(providerApiKey('api.wuhen-ai.com')),
    balance: '', unit: '', isValid: false, planName: '',
    today: { requests: 0, tokens: 0, cost: 0 },
    total: { requests: 0, tokens: 0, cost: 0 },
    daily: [], models: [],
    error: '', updatedAt: new Date().toISOString(),
  };
}

function fetchUsage(apiKey) {
  return new Promise((resolve) => {
    const req = https.request(USAGE_URL, {
      method: 'GET',
      timeout: 3000,
      headers: { Authorization: `Bearer ${apiKey}`, Accept: 'application/json' },
    }, (res) => {
      let data = '';
      res.on('data', (c) => { data += c; });
      res.on('end', () => {
        try {
          const obj = JSON.parse(data);
          if (res.statusCode !== 200) return resolve(null);
          resolve(obj);
        } catch {
          resolve(null);
        }
      });
    });
    req.on('timeout', () => { req.destroy(); resolve(null); });
    req.on('error', () => resolve(null));
    req.end();
  });
}

async function refreshOnce() {
  const apiKey = providerApiKey('api.wuhen-ai.com');
  if (!apiKey) {
    cache = {
      ts: Date.now(),
      payload: {
        enabled: false, balance: '', unit: '', isValid: false, planName: '',
        today: { requests: 0, tokens: 0, cost: 0 },
        total: { requests: 0, tokens: 0, cost: 0 },
        daily: [], models: [],
        error: '', updatedAt: new Date().toISOString(),
      },
    };
    return;
  }

  const obj = await fetchUsage(apiKey);
  if (!obj) {
    // 失败：复用旧缓存（若有效）；否则写 error 结构
    if (cache.payload && cache.payload.enabled && cache.payload.balance !== '') return;
    cache = {
      ts: Date.now(),
      payload: {
        enabled: true, balance: '', unit: '', isValid: false, planName: '',
        today: { requests: 0, tokens: 0, cost: 0 },
        total: { requests: 0, tokens: 0, cost: 0 },
        daily: [], models: [],
        error: '查询失败', updatedAt: new Date().toISOString(),
      },
    };
    return;
  }

  const today = obj.usage && obj.usage.today || {};
  const total = obj.usage && obj.usage.total || {};
  cache = {
    ts: Date.now(),
    payload: {
      enabled: true,
      balance: obj.balance ?? obj.remaining ?? '',
      unit: obj.unit || 'USD',
      isValid: obj.isValid ?? true,
      planName: obj.planName || '',
      today: {
        requests: today.requests || 0,
        tokens: today.total_tokens || 0,
        cost: today.cost || 0,
        actualCost: today.actual_cost || 0,
      },
      total: {
        requests: total.requests || 0,
        tokens: total.total_tokens || 0,
        cost: total.cost || 0,
        actualCost: total.actual_cost || 0,
      },
      // 近 7 天按日（旧日期在前，取尾部翻转成近到远）
      daily: (obj.daily_usage || []).slice(-7).map((d) => ({
        date: d.date,
        requests: d.requests || 0,
        tokens: d.total_tokens || 0,
        cost: d.cost || 0,
      })),
      models: (obj.model_stats || []).map((m) => ({
        model: m.model || '',
        requests: m.requests || 0,
        tokens: m.total_tokens || 0,
        cost: m.cost || 0,
      })),
      error: '',
      updatedAt: new Date().toISOString(),
    },
  };
}

module.exports = { refreshOnce, getWuhenUsage, CACHE_TTL };

File: src-tauri/src/data/wuhen.rs
// 无痕中转（api.wuhen-ai.com）— 余额 + 用量查询，对齐 Electron src/data/wuhen.js。
// key 从 ~/.zcode/v2/config.json 按 baseURL 关键字匹配；GET /v1/usage 一次拿
// 余额 + 今日/累计 + 按日 + 按模型。后台刷新 + 缓存。

use std::sync::{Mutex, OnceLock};

use serde_json::{json, Value};

use crate::data::volc::provider_api_key;
use crate::data::Cache;

pub const CACHE_TTL: u64 = 15;
const USAGE_URL: &str = "https://api.wuhen-ai.com/v1/usage";

fn usage_cache() -> &'static Mutex<Cache> {
    static CACHE: OnceLock<Mutex<Cache>> = OnceLock::new();
    CACHE.get_or_init(|| Mutex::new(Cache::new()))
}

fn empty_payload(enabled: bool, error: &str) -> Value {
    json!({
        "enabled": enabled,
        "balance": "",
        "unit": "",
        "isValid": false,
        "planName": "",
        "today": { "requests": 0, "tokens": 0, "cost": 0 },
        "total": { "requests": 0, "tokens": 0, "cost": 0 },
        "daily": [],
        "models": [],
        "error": error,
        "updatedAt": chrono::Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Millis, true),
    })
}

pub fn get_wuhen_usage() -> Value {
    let cached = usage_cache().lock().map(|c| c.get()).unwrap_or(Value::Null);
    if cached.is_null() {
        empty_payload(!provider_api_key("api.wuhen-ai.com").is_empty(), "")
    } else {
        cached
    }
}

async fn fetch_usage(api_key: &str) -> Option<Value> {
    let body = crate::data::net::get(USAGE_URL, &[("Authorization", &format!("Bearer {api_key}")), ("Accept", "application/json")]).await?;
    serde_json::from_str(&body).ok()
}

pub async fn refresh_once() {
    let api_key = provider_api_key("api.wuhen-ai.com");
    if api_key.is_empty() {
        usage_cache().lock().unwrap().set(empty_payload(false, ""));
        return;
    }

    // 网络调用期间不持锁（MutexGuard 跨 await 会让 future 非 Send）
    let obj = fetch_usage(&api_key).await;
    let mut guard = usage_cache().lock().unwrap();
    let Some(obj) = obj else {
        // 失败：复用旧缓存（若有效）；否则写 error 结构
        let old = guard.get();
        let has_valid = old.get("enabled").and_then(|v| v.as_bool()).unwrap_or(false)
            && old.get("balance").and_then(|v| v.as_str()).map(|s| !s.is_empty()).unwrap_or(false);
        if !has_valid {
            guard.set(empty_payload(true, "查询失败"));
        }
        return;
    };

    let usage = obj.get("usage").cloned().unwrap_or(json!({}));
    let today = usage.get("today").cloned().unwrap_or(json!({}));
    let total = usage.get("total").cloned().unwrap_or(json!({}));

    let daily: Vec<Value> = obj
        .get("daily_usage")
        .and_then(|v| v.as_array())
        .map(|arr| {
            arr.iter()
                .rev()
                .take(7)
                .map(|d| {
                    json!({
                        "date": d.get("date").cloned().unwrap_or(json!("")),
                        "requests": d.get("requests").and_then(|v| v.as_i64()).unwrap_or(0),
                        "tokens": d.get("total_tokens").and_then(|v| v.as_i64()).unwrap_or(0),
                        "cost": d.get("cost").and_then(|v| v.as_i64()).unwrap_or(0),
                    })
                })
                .collect::<Vec<Value>>()
        })
        .unwrap_or_default()
        .into_iter()
        .rev()
        .collect();

    let models: Vec<Value> = obj
        .get("model_stats")
        .and_then(|v| v.as_array())
        .map(|arr| {
            arr.iter()
                .map(|m| {
                    json!({
                        "model": m.get("model").cloned().unwrap_or(json!("")),
                        "requests": m.get("requests").and_then(|v| v.as_i64()).unwrap_or(0),
                        "tokens": m.get("total_tokens").and_then(|v| v.as_i64()).unwrap_or(0),
                        "cost": m.get("cost").and_then(|v| v.as_i64()).unwrap_or(0),
                    })
                })
                .collect()
        })
        .unwrap_or_default();

    guard.set(json!({
        "enabled": true,
        "balance": obj.get("balance").or_else(|| obj.get("remaining")).cloned().unwrap_or(json!("")),
        "unit": obj.get("unit").and_then(|v| v.as_str()).unwrap_or("USD"),
        "isValid": obj.get("isValid").and_then(|v| v.as_bool()).unwrap_or(true),
        "planName": obj.get("planName").and_then(|v| v.as_str()).unwrap_or(""),
        "today": {
            "requests": today.get("requests").and_then(|v| v.as_i64()).unwrap_or(0),
            "tokens": today.get("total_tokens").and_then(|v| v.as_i64()).unwrap_or(0),
            "cost": today.get("cost").and_then(|v| v.as_i64()).unwrap_or(0),
            "actualCost": today.get("actual_cost").and_then(|v| v.as_i64()).unwrap_or(0),
        },
        "total": {
            "requests": total.get("requests").and_then(|v| v.as_i64()).unwrap_or(0),
            "tokens": total.get("total_tokens").and_then(|v| v.as_i64()).unwrap_or(0),
            "cost": total.get("cost").and_then(|v| v.as_i64()).unwrap_or(0),
            "actualCost": total.get("actual_cost").and_then(|v| v.as_i64()).unwrap_or(0),
        },
        "daily": daily,
        "models": models,
        "error": "",
        "updatedAt": chrono::Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Millis, true),
    }));
}
